src/PersistentOrderedMap.mo

```diff
@@ -126,258 +147,422 @@ module {
   /// where `n` denotes the number of key-value entries stored in the tree.
   ///
   /// Note: Creates `O(log(n))` temporary objects that will be collected as garbage.
-  public func size<X, Y>(t : Map<X, Y>) : Nat {
+  public func size<K, V>(t : Map<K, V>) : Nat {
     switch t {
       case (#leaf) { 0 };
-      case (#node(_, l, xy, r)) {
-        size(l) + size(r) + (switch (xy.1) { case null 0; case _ 1 })
+      case (#node(_, l, _, r)) {
+        size(l) + size(r) + 1
       }
     }
   };
 
-  func redden<X, Y>(t : Map<X, Y>) : Map<X, Y> {
-    switch t {
-      case (#node (#B, l, xy, r)) {
-        (#node (#R, l, xy, r))
+  func iterStack<T>(stack : S.Stack<T>) : I.Iter<T> = { next = stack.pop };
+
+  public func foldLeft<Key, Value, Accum>(
+    step : (Key, Value, Accum) -> Accum,
+    empty : Accum,
+    map : Map<Key, Value>
+  ) : Accum
+  {
+    func getMin(
+      stack : Internal.MapStack<Key, Value>,
+      folder : Internal.Folder<Key, Value, Accum>,
+      acc : Accum
+    ) : Accum
+    {
+      for (tree in iterStack(stack)) {
+        switch(tree) {
+          case (#leaf) { };
+          case (#node(_, #leaf, val, b)) {
+            Internal.pushMap(b, stack);
+            return (folder val)
+          };
+          case (#node(_, a, val, b)) {
+            Internal.pushNode(a, val, b, stack);
+          }
+        }
       };
-      case _ {
-        Debug.trap "RBTree.red"
-      }
-    }
+      acc
+    };
+    Internal.fold(getMin, step, empty, map)
   };
 
-  func lbalance<X,Y>(left : Map<X, Y>, xy : (X,?Y), right : Map<X, Y>) : Map<X,Y> {
-    switch (left, right) {
-      case (#node(#R, #node(#R, l1, xy1, r1), xy2, r2), r) {
-        #node(
-          #R,
-          #node(#B, l1, xy1, r1),
-          xy2,
-          #node(#B, r2, xy, r))
-      };
-      case (#node(#R, l1, xy1, #node(#R, l2, xy2, r2)), r) {
-        #node(
-          #R,
-          #node(#B, l1, xy1, l2),
-          xy2,
-          #node(#B, r2, xy, r))
+  public func foldRight<Key, Value, Accum>(
+    step : (Key, Value, Accum) -> Accum,
+    empty : Accum,
+    map : Map<Key, Value>
+  ) : Accum
+  {
+    func getMax(
+      stack : Internal.MapStack<Key, Value>,
+      folder : Internal.Folder<Key, Value, Accum>,
+      acc : Accum
+    ) : Accum
+    {
+      for (tree in iterStack(stack)) {
+        switch(tree) {
+          case (#leaf) { };
+          case (#node(_, a, val, #leaf)) {
+            Internal.pushMap(a, stack);
+            return (folder val);
+          };
+          case (#node(_, a, val, b)) {
+            Internal.pushNode(b, val, a, stack);
+          }
+        }
       };
```

Comment:
Shouldn't it be just a helper function in the `Internal` module? I don't see any closure here.

Reply:
It captures types (Key, Value, Accum), not terms

It is possible to write
```motoko
    public func getMin<Key, Value, Accum>(
      stack : Internal.MapStack<Key, Value>,
      folder : Internal.Folder<Key, Value, Accum>,
      acc : Accum
    ) : Accum
```
But then you can't use it like this:
```motoko
Internal.fold(Internal.getMin, step, empty, map)
```
You will get the following type error then:
<details>
<summary>Big error message</summary>

```
cannot implicitly instantiate function of type
  <Key, Value, Accum>(FoldGetter<Key, Value, Accum>,
                      (Key, Value, Accum) -> Accum, Accum, Map<Key, Value>) ->
    Accum
to argument of type
  (<Key_1, Value_1, Accum_1>(MapStack<Key_1, Value_1>,
                             Folder<Key_1, Value_1, Accum_1>, Accum_1) ->
     Accum_1,
   (Key__4, Value__4, Accum__4) -> Accum__4, Accum__4,
   {
     #leaf;
     #node :
       (Color, Map<Key__4, Value__4>, (Key__4, Value__4),
        Map<Key__4, Value__4>)
   })
to produce result of type
  Accum__4
because no instantiation of Key__9, Value__9, Accum__9 makes
  (<Key_1, Value_1, Accum_1>(MapStack<Key_1, Value_1>,
                             Folder<Key_1, Value_1, Accum_1>, Accum_1) ->
     Accum_1,
   (Key__4, Value__4, Accum__4) -> Accum__4, Accum__4,
   {
     #leaf;
     #node :
       (Color, Map<Key__4, Value__4>, (Key__4, Value__4),
        Map<Key__4, Value__4>)
   })  <:
    (FoldGetter<Key__9, Value__9, Accum__9>,
     (Key__9, Value__9, Accum__9) -> Accum__9, Accum__9,
     Map<Key__9, Value__9>)
and
  Accum__9  <:  Accum__4
```
</details>

You cannot instantiate `getMin` without applying it to the arguments and supplying exact type information. We can wrap it manually, though:
```motoko
  public func foldLeft<Key, Value, Accum>(
    step : (Key, Value, Accum) -> Accum,
    empty : Accum,
    map : Map<Key, Value>
  ) : Accum
  {
    let getMin : Internal.FoldGetter<Key, Value, Accum>
    = func (stack,folder,acc) {Internal.getMin(stack,folder,acc)};

    Internal.fold(getMin, step, empty, map)
  };
```
I like this solution because we don't have to export internal helpers from, well, Internal module, but the wrapping is a bit annoying.

Should we proceed this way?